store/src/event.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashMap;
+
+use anyhow::{anyhow, Context};
+use ceramic_core::{EventId, Network};
+use ceramic_event::unvalidated;
+use cid::Cid;
+use iroh_car::CarReader;
+use tokio::io::AsyncRead;
+
+use crate::{Error, Result};
+
+// Helper function to construct an event Id from CAR data of an event
+// We should likely move this closer to where its needed but this is good enough for now.
+pub(crate) async fn event_id_from_car<R>(network: Network, reader: R) -> Result<EventId>
```

Comment:
The CAR file contains other blocks that we want to parse in order to get all the data. So having access to the entire CAR file is important as part of the process of constructing the EventID. 

I briefly considered using an event based parser (i.e. SAX XML parser) and then let there be various consumers of the parser events to build up the state it needs. However that felt overly complex for the needs here as its easy to ensure we only read/parse a block once.

Reply:
huh?  I don't understand your concern about needing the CAR file.  My suggestion was that we fully parse everything out of the CAR file and into an Event struct with everything we'll ever need to access for the Event, and that that Event struct has a `get_event_id()` method on it.

Reply:
Constructing an event ID from an event has two dependencies. 

1. The blocks in the CAR file
2. The init block in the store

The CAR file for the simplest case has two blocks that we need, the envelope and the payload. The Event type we have created represents the raw structure of the IPLD data. Specifically the Event enum represents the envelope while the Payload enum represents the payload block. We do not have a reference from the Event enum to the Payload enum. They are loosely connected via the CID in the envelope.

 In order the construct the event ID from an update event for example we do the following steps:
 
 1. Decode the CAR file and get the CID of the first root.
 2. Decode the block from the CAR for that CID into an Event (i.e. envelope)
 3. Get the CID of the payload from the of the Event (envelope)
 4. Decode the payload block from the CAR 
 5. Get the CID of the init event from the payload
 6. Read the init event from the store and decode as an Event
 7. Get the payload CID from the init Event
 8. Read the payload block from the store and decode
 9. Get the controller and model from the init payload header
 
 These steps require database reads and access to the CAR file in order to read blocks that exist in the CAR skipping a database read.
 
 Given that the Event type and its friends are data structs, meaning simply contain deserialzed data and do not contain references to each other it makes sense to me to have a function not attached to the Event type that can do this operation when provided with the dependencies.
 
 >  My suggestion was that we fully parse everything out of the CAR file and into an Event struct with everything we'll ever need to access for the Event
 
 How would you do this? Does this imply that we treat CIDs as references ? If so does that mean you can't construct an event without materializing the entire log of events back to init for that event since that is where the trail of references end? 
 
 Happy to discuss on a call if I am not making sense. TL;DR I want to decouple these data structure from how they are used because they can be used in many contexts which may not all need to do the same operations and therefore do not have the same dependencies.

Reply:
I wouldn't imagine materializing the entire log, just all the blocks that are a part of this event (so following all links *except* for the `prev` link), plus any extra metadata we need such as the controller and model.

As discussed on discord though, happy to leave this for a future optimization.